cli/alert.go

```diff
@@ -0,0 +1,189 @@
+	Long: `View and search through current alerts.
+
+  Amtool has a simplified prometheus query syntax, but contains robust support for
+  bash variable expansions. The non-option section of arguments constructs a list
+  of "Matcher Groups" that will be used to filter your query. The following
+  examples will attempt to show this behaviour in action:
+
+  amtool alert query alertname=foo node=bar
+
+  	This query will match all alerts with the alertname=foo and node=bar label
+  	value pairs set.
+
+  amtool alert query foo node=bar
+
+  	If alertname is ommited and the first argument does not contain a '=' or a
+  	'=~' then it will be assumed to be the value of the alertname pair.
+
+  amtool alert query 'alertname=~foo.*'
+
+  	As well as direct equality, regex matching is also supported. The '=~' syntax
+  	(similar to prometheus) is used to represent a regex match. Regex matching
+  	can be used in combination with a direct match.
+
+  amtool alert query alertname=foo node={bar,baz}
+
+  	This query will match all alerts with the alertname=foo label value pair
+  	and EITHER node=bar or node=baz.
+
+  amtool alert query alertname=foo{a,b} node={bar,baz}
+
+	Similar to the previous example this query will match all alerts with any
+	combination of alertname=fooa or alertname=foob AND node=bar or node=baz.
+
+	`,
+	RunE: queryAlerts,
+}
+
+func init() {
+	RootCmd.AddCommand(alertCmd)
+	alertCmd.Flags().Bool("expired", false, "Show expired alerts as well as active")
+	alertCmd.Flags().BoolP("silenced", "s", false, "Show silenced alerts")
+	alertFlags = alertCmd.Flags()
+}
+
+func fetchAlerts(filter *[]labels.Matcher) (model.Alerts, error) {
+	alertResponse := alertmanagerAlertResponse{}
+
+	u, err := GetAlertmanagerURL()
+	if err != nil {
+		return model.Alerts{}, err
+	}
+
+	u.Path = path.Join(u.Path, "/api/v1/alerts")
+	if filter != nil {
+		u.RawQuery = url.QueryEscape(MatchersToString(*filter))
+	}
+
+	fmt.Println(u.String())
+
+	res, err := http.Get(u.String())
+	if err != nil {
+		return model.Alerts{}, err
+	}
+
+	defer res.Body.Close()
+	decoder := json.NewDecoder(res.Body)
+
+	err = decoder.Decode(&alertResponse)
+	if err != nil {
+		return model.Alerts{}, errors.New("Unable to decode json response")
+	}
+	return alertResponse.Data, nil
+}
+
+func queryAlerts(cmd *cobra.Command, args []string) error {
+	silences, err := fetchSilences(nil)
+	if err != nil {
+		return err
+	}
+
+	var groups [][]labels.Matcher
+	if len(args) > 0 {
+		matchers, err := parseMatchers(args)
+		if err != nil {
+			return err
+		}
+		groups = parseMatcherGroups(matchers)
+		if err != nil {
+			return err
+		}
+	}
+
+	expired, err := alertFlags.GetBool("expired")
+	if err != nil {
+		return err
+	}
+
+	showSilenced, err := alertFlags.GetBool("silenced")
+	if err != nil {
+		return err
+	}
+
+	fetchedAlerts := model.Alerts{}
+	if len(groups) < 1 {
+		fmt.Println("No specified matchers")
+		fetchedAlerts, err = fetchAlerts(nil)
```

Comment:
A slice contains a pointer to an internal array, so passing around a pointer to a slice isn't really something that happens in go.

The flow of this seems to be doing a lot of work:

1. Parse the user input into matchers
2. Pass matchers to fetch function
3. Stringify matchers for the query string

I think it would make sense to massage the string args into the format that gets sent to the server, and then let the server handle parsing and error handling:

```go
l := fmt.Sprintf("{%s}", strings.Join(os.Args[2:], ","))
fetched, err := fetchAlerts(l)

// in fetchAlerts
qs := url.QueryEscape("filter=" + l)
```

And just send that to the back end -- if there's an error, it'll be handled consistently from the server.